### ISSUE TITLE
Notify user code about contexts created in the background

### ISSUE THREAD

**Issue by JoniSt** [HappySeaFox/sail#97]
I'm currently using SAIL in a heavily multithreaded project where I don't know exactly what threads my code (and therefore SAIL) is potentially running on. This means that SAIL will often create new thread-local contexts in the background that my code isn't aware of (and therefore can't clean up with `sail_finish`).

I'd really appreciate it if you could add a feature that allows user code to get notified of new contexts created by SAIL. [Here's a patch](https://gist.github.com/JoniSt/bec52ad0a2f571006a047dcb98e57865) that implements this functionality with a user-settable callback function which gets called by `init_context` just before context initialization. That way, the user can manipulate the flags of the context and keep track of it so it can be properly cleaned up when it's not needed anymore (i.e. when the thread exits).

Thanks!

**Comment by HappySeaFox** [comment]
Thanks, @JoniSt ! I'll review the patch and the request when I'm back online.

**Comment by HappySeaFox** [comment]
Thanks again for the request.

I took some time to think about it. Currently, to support multi-threaded applications, I use a pretty naive approach to initialize a new context with a independent set of codecs per thread. Not the best idea for performance reasons and we have a problem to cleanup contexts after a thread exits.

Currently I'm thinking about having just a single global context. It will be initialized just once. Mutexes will be used to guard it internally to support multi-threaded applications. Decoding and encoding functions are already thread-safe per se.

PROS:
- No need to cleanup unused contexts when threads exit.
- Initialize context just once. Other threads will simply re-use the already existing context and codecs. Currently, every thread has an independent copy of codecs.

CONS:
- Performance penalty is that we will lock a mutex **every time** we need to start decoding an image. When the context is already initialized, it will waste a "couple" of CPU cycles to go to the kernel if the mutex is locked already by a different thread. Not a big deal for me comparing to having an independent set of codecs per thread as currently implemented.

Please share your ideas.

**Comment by JoniSt** [comment]
That would work quite nicely too!

Given that SAIL currently loads all codecs at once, it might also be possible to avoid locking the mutex when the context already exists (using double-checked locking).

Something like this:
```
sail_status_t current_global_context_with_flags(struct sail_context **context, int flags) {
    SAIL_CHECK_CONTEXT_PTR(context);
    if (!atomic_load(g_sail_global_context)) {
        mutex_lock(g_sail_global_context_mutex);
        if (!g_sail_global_context) {
            SAIL_TRY_OR_CLEANUP(alloc_context(context), mutex_unlock(g_sail_global_context_mutex));
            SAIL_TRY_OR_CLEANUP(init_context(*context, flags), mutex_unlock(g_sail_global_context_mutex));
            atomic_store(&g_sail_global_context, *context);
        }
        mutex_unlock(g_sail_global_context_mutex);
    }
    *context = g_sail_global_context;
    return SAIL_OK;
}
```

That way, the mutex gets ignored entirely once SAIL has loaded the codecs. The only synchronization required is an atomic load, which is super cheap (just a normal memory load on x86). However, this also means that `sail_finish` is not thread-safe anymore as it might destroy the context while other threads still need it. So the user has to make sure to only call it once SAIL is no longer needed by any thread. The context flags might also need to be stored in a global variable (so the user can set them at program startup with something like `sail_set_default_context_flags`).

**Comment by HappySeaFox** [comment]
@JoniSt Thanks for the input!

>something like sail_set_default_context_flags

This one already exists - `sail_init_with_flags()`. If you want to initialize context explicitly, you can use this function.

Thanks for the code sample as well, I'll look into refactoring `libsail` to use a single global context.

**Comment by HappySeaFox** [comment]
@JoniSt Could you please try from the branch `feature/use-single-global-context`? If you use VCPKG, you can try the following patch: [vcpkg-use-global-context.diff.txt](https://github.com/HappySeaFox/sail/files/7007106/vcpkg-use-global-context.diff.txt)

**Comment by JoniSt** [comment]
I've just updated my codebase to use this new version of SAIL. Everything works great and the mutex guards look good to me. Thank you very much!

I've tested it both on Linux x64 (Debian, GCC 10.2) as well as on Windows 10 x64 (MinGW, also GCC 10.2).

The time spent holding the mutex is also quite small compared to the time needed to decode an image, so threads shouldn't fight for that mutex much - while one thread is decoding an image, another thread can get the mutex and fetch its codec from the global state. So this all works out very nicely.
